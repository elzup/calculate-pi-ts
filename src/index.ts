const multiple = (a: number, b: number) => a * b;
const range0 = (n: number) => [...Array(n).keys()];
const range = (n: number) => {
  const r = range0(n + 1);
  r.shift();
  return r;
};
const factorial = (n: number) => range(n).reduce(multiple, 1);

const memo: { [n: number]: number } = {};
const factorialMemo = (n: number) =>
  n in memo ? memo[n] : (memo[n] = factorial(n));

const A = 13591409;
const B = 545140134;
const C = 640320 ** 3;

const chudnovskySig = (n: number) =>
  ((-1) ** n * factorialMemo(6 * n) * (A + B * n)) /
  (factorialMemo(3 * n) * factorialMemo(n) ** 3 * C ** n);

export function chudnovsky(t: number) {
  const res =
    (12 / Math.sqrt(C)) * range0(t).reduce((p, i) => p + chudnovskySig(i), 0);
  return 1 / res;
}
