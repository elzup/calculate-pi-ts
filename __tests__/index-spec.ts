import { chudnovsky } from "../src";

test("Should have Greeter available", () => {
  expect(chudnovsky(10)).toMatchSnapshot();
});
