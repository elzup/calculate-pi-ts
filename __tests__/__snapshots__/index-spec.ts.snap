// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should have Greeter available 1`] = `3.1415926535897936`;
